core/dbt/graph/graph.py

```diff
@@ -33,11 +33,23 @@ def ancestors(self, node: UniqueId, max_depth: Optional[int] = None) -> Set[Uniq
             for _, child in nx.bfs_edges(self.graph, node, reverse=True, depth_limit=max_depth)
         }
 
-    def descendants(self, node: UniqueId, max_depth: Optional[int] = None) -> Set[UniqueId]:
+    def descendants(
+        self, node: UniqueId, max_depth: Optional[int] = None, include_tests_children: bool = False
+    ) -> Set[UniqueId]:
         """Returns all nodes reachable from `node` in `graph`"""
         if not self.graph.has_node(node):
             raise DbtInternalError(f"Node {node} not found in the graph!")
-        return {child for _, child in nx.bfs_edges(self.graph, node, depth_limit=max_depth)}
+        filtered_graph = nx.restricted_view(
```

Comment:
Good news! If we apply this exact same filtering logic to `ancestors` above, we could also resolve a similar bug reported for upstream selection (https://github.com/dbt-labs/dbt-core/issues/5193).

@b-luu Is that something you'd be willing to take on, in the scope of this same PR? I think it would just require factoring this out into a reusable method, and one more (slightly more involved) test case, asserting that `dbt ls 1+model_c` and `dbt build 1+model_c` return the same number of results.

Reply:
Oh yeah sure: 2 birds in one stone 👍

I'll factor that in during the week (probably Wednesday).

Reply:
Thanks for spotting that!
I had thought about adding it more generally just in case, but didn't see that particular corner case that really justifies it.

Come to think of it, I think these added test edges should simply be ignored from the whole selection process in itself. I'll try to frame it in more generally if I can.

Reply:
Turns out I couldn't sleep this morning, so here it is ;)

Reply:
I limited myself to fitting the new method in ancestors and descendants because I didn't have time to thoroughly dig through and test all the other graph methods to see if we could/should expand this filtering further... :shrug: 

But just covering ancestors and children is enough to solve both issues listed :+1: